enhancements/baremetal/an-slo-for-baremetal.md

```diff
@@ -0,0 +1,432 @@
+  `Deployment`, copying much of the MAO pod spec in terms of
+  `system-node-critical` priority class, running on masters, security
+  context, resource requests, etc.
+- Implement a controller reconciling the singleton
+  `provisioning-configuration` cluster-scoped `Provisioning` resource
+- Do nothing except set `Disabled=true`, `Available=true`, and
+  `Progressing=False` when the `Infrastructure` resource a platform
+  type other than `BareMetal`.
+- Based on the values in the `Provisioning` resource, create a
+  `metal3` `Deployment` and associated `metal3-mariadb-password` under
+  the `openshift-machine-api` namespace. This is the same as the MAO
+  currently creates.
+
+### Test Plan
+
+### Graduation Criteria
+
+### Upgrade / Downgrade / Verson Skew Strategy
+
+The key thing to consider for upgrades is how to smoothly transition
+the `metal3` deployment in the `openshift-machine-api` namespace from
+the MAO to the CBO.
+
+We assume that the CBO will first appear in `4.N.0` and take over from
+the MAO in that release. We can merge changes into `4.N-1.y` paving
+the way for this and, if necessary, require an upgrade to given
+`4.N-1.y` version before the upgrade to `4.N.0`. Since we only support
+upgrading directly from `4.N-1` to `4.N`, we can remove all upgrade
+handling code in `4.N+1`.
+
+#### Prior Art
+
+We will take inspiration from the
+[Separate OAuth API-Resources](https://github.com/openshift/enhancements/blob/master/enhancements/authentication/separate-oauth-resources.md)
+enhancement
+
+> Add code in 4.(n-1) to the cluster-openshift-apiserver-operator that
+> prevents it from managing an apiservice if it is "claimed" by the
+> cluster-authentication-operator via an annotation and if the
+> cluster-authentication-operator is at least at 4.n.
+
+In the related
+[cluster-openshift-apiserver-operator PR](https://github.com/openshift/cluster-openshift-apiserver-operator/pull/294)
+the details are clear - `cluster-authentication-operator` indicates
+it is "at least at 4.n" by setting the `managingOAuthAPIServer` to
+`True` in the status field of its. Then `cluster-authentication-operator`
+"claims" an apiservice by setting the `authentication.operator.openshift.io/managed`
+annotation.
+
+Presumably the "at least at 4.n" check is required in addition to the
+annotation to handle a downgrade scenario - where the annotation is
+set, but `cluster-authentication-operator` has been downgraded.
+
+#### Design
+
+We need the following changes to MAO:
+
+1. Respect a "gate" allowing CBO "claim" the resource.
+
+   We will use [the existing `machine.openshift.io/owned` annotation](http://github.com/openshift/machine-api-operator/pull/424)
+   on the `metal3` deployment to indicate that MAO is managing the
+   resource. A new `baremetal.openshift.io/owned` annotation will
+   indicate that CBO is in control. Only one of these annotations
+   should be set on a resource but, for the avoidance of doubt, CBO
+   takes precedence.
```

Comment:
> I see that MAO opted to [not use an owner reference](https://github.com/openshift/machine-api-operator/pull/424#issuecomment-544895376) for reasons specific to them. From what I understand of this proposal, CBO would not have such a limitation.

I assume you're suggesting that the CBO deployment would be what we'd list as the owner of the metal3 deployment?

The logic appears to apply here? Is running MAO off-cluster any more of a valid use case than running CBO off-cluster? The only use case I think of is during development?

And likewise, "the controller itself is the owner" comment seems to apply here. Unless it could make sense to make the provisioning resource the owner?

(I'm not sure I can judge how valid the MAO justification is ... I'm just saying it appears to apply here too)

> Since CBO is taking over as the long-term owner, it's worth considering whether it makes sense for it to express its ownership through a standard owner reference instead of an annotation. It could do both, but that diminishes the value of the annotation.
> 
> There is value in the consistency of having both operators use an annotation, though they each have their own annotation and thus their own independent ways of identifying their ownership. Maybe it makes more sense to optimize for the long-term state than the transitional state, and not have a `baremetal.openshift.io/owned` annotation to either get rid of later or retain as technical debt.

It's an interesting suggestion. I must admit though I find it challenging to think through. There are 2 aspects:

1) What would be an appropriate ownerRef for the `metal3` deployment?

I struggle with this one because I feel like there's tons of "case law" I'm missing.

For example, looking at my running 4.5 cluster, I see exactly 1 owner reference on a deployment - the OLM packageserver deployment lists its ClusterServiceVersion as an owner.

2) Are there any gotchas with using ownerReferences for the purposes of this upgrade hand-off?

I can't think of any particularly, but I worry I'm missing something.

Here's my first cut at code in MAO to stand down if the CBO has claimed ownership: https://github.com/openshift/machine-api-operator/compare/master...markmc:cbo-gate